src/test/java/org/threeten/extra/TestYearHalf.java

```diff
@@ -0,0 +1,999 @@
+                            assertEquals(true, b.isAfter(a));
+                            assertEquals(true, a.isBefore(b));
+                        } else if (year1 > year2) {
+                            assertEquals(true, a.compareTo(b) > 0);
+                            assertEquals(true, b.compareTo(a) < 0);
+                            assertEquals(true, a.isAfter(b));
+                            assertEquals(true, b.isBefore(a));
+                            assertEquals(false, b.isAfter(a));
+                            assertEquals(false, a.isBefore(b));
+                        } else {
+                            if (half1.getValue() < half2.getValue()) {
+                                assertEquals(true, a.compareTo(b) < 0);
+                                assertEquals(true, b.compareTo(a) > 0);
+                                assertEquals(false, a.isAfter(b));
+                                assertEquals(false, b.isBefore(a));
+                                assertEquals(true, b.isAfter(a));
+                                assertEquals(true, a.isBefore(b));
+                            } else if (half1.getValue() > half2.getValue()) {
+                                assertEquals(true, a.compareTo(b) > 0);
+                                assertEquals(true, b.compareTo(a) < 0);
+                                assertEquals(true, a.isAfter(b));
+                                assertEquals(true, b.isBefore(a));
+                                assertEquals(false, b.isAfter(a));
+                                assertEquals(false, a.isBefore(b));
+                            } else {
+                                assertEquals(0, a.compareTo(b));
+                                assertEquals(0, b.compareTo(a));
+                                assertEquals(false, a.isAfter(b));
+                                assertEquals(false, b.isBefore(a));
+                                assertEquals(false, b.isAfter(a));
+                                assertEquals(false, a.isBefore(b));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    public void test_compareTo_nullYearHalf() {
+        assertThrows(NullPointerException.class, () -> TEST.compareTo(null));
+    }
+
+    //-----------------------------------------------------------------------
+    // equals() / hashCode()
+    //-----------------------------------------------------------------------
+    @Test
+    public void test_equals() {
+        for (int year1 = -100; year1 < 100; year1++) {
+            for (Half half1 : Half.values()) {
+                YearHalf a = YearHalf.of(year1, half1);
+                for (int year2 = -100; year2 < 100; year2++) {
+                    for (Half half2 : Half.values()) {
+                        YearHalf b = YearHalf.of(year2, half2);
+                        if (year1 == year2 && half1 == half2) {
```

Comment:
Wouldn't something like this be better? `a` is not really "expected" value in this context.

```java
if (year1 == year2 && half1 == half2) {
    assertTrue(a.equals(b));
    assertTrue(a.hashCode() == b.hashCode());
} else {
    assertFalse(a.equals(b));
    // can _not_ assert about hashCode
}
```
I've added `assertFalse` as a bonus. Before addition `equals` could always returns `true` and the test would still pass.

Reply:
Using the "normal" `assertEquals` means that if the test fails the output shows both sides of the failure. If you use `assertTrue` you have to manually rewrite or add logging to discover what is wrong. On a project like this, where CI runs in the cloud, the `assertEquals` way works out better.

Reply:
`assertEquals` _requires_ `equals` to be written properly, otherwise you get nonsense in your logs or no logs at all. Imagine `equals` always returning `true`: you will get nothing logged with assertions like `assertEquals(a, b)`. Imagine `equals` always returning `false`: you will get bunch of `"expected x and got x". Logs like this are useless.

`assertEquals` is clearly not meant for testing `equals`. Testing `equals` should involve direct checking whether `true` or `false` is being returned for specific pair of arguments. `assertEquals` shines with `assertEquals("2001-H2", "2001-H1".plusHalfs(1))` and similar.

If you are concerned about CI and logging, maybe you should switch to [ParametrizedTest](https://junit.org/junit5/docs/current/api/org.junit.jupiter.params/org/junit/jupiter/params/ParameterizedTest.html)s? This way you will know exactly, what went wrong (and not: what's the first thing, that went wrong). It that's the case, I'll be happy to help.

**You still do not test at all if `a.equals(b)` ever returns something other than `true`.**

Reply:
If `@ParametrizedTest`s is too much, then there is always possibility to add custom message (lazy or not), like
```java
assertTrue(a.equals(b), () -> a + " .equals " + b);
```
which gives exceptions similar to
`org.opentest4j.AssertionFailedError: 2001-H1 .equals 2001-H1 ==> expected: <true> but was: <false>`

Reply:
See https://github.com/ThreeTen/threeten-extra/pull/315